Texture filtering and wrapping

aferriss opened the pull request:
I was wanting some more control over how textures were filtered and wrapped so I implemented these two functions on top of the Texture class. 

````
tex.setInterpolation( downscale, upscale);
// downscale & upscale can be constants NEAREST or LINEAR

tex.setWrapMode( wrapX, wrapY);
// wrapX and wrapY can be CLAMP, REPEAT, or MIRROR
// if the texture is not power of two it will fallback to clamp
````

WebGL2 does support non power of two texture for REPEAT and MIRROR but since there wasn't an obvious way to get at version information I just implemented it as a webGL1 function. I also left out mipmapping for now.

You can use them like so...
````
let canvas, img;
function preload(){
   img = loadImage('img.jpg');
}

function setup(){
  canvas = createCanvas(400,400,WEBGL);
  let tex = canvas.getTexture(img);
  tex.setTextureWrap(REPEAT, REPEAT);
  tex.setInterpolation(NEAREST, NEAREST);
}

````

kjhollen commented on src/webgl/p5.Texture.js:
```diff
@@ -252,4 +253,112 @@ p5.Texture.prototype.unbindTexture = function() {
+    this.glMinFilter = gl.LINEAR;
+  } else {
+    // falling back to default
+    this.glMinFilter = gl.LINEAR;
```
Is the format of this block in anticipation of adding more possible scaling constants in the future? both of the last two conditions result in gl.LINEAR (and same is done below).

aferriss replied:
I think I did this on the off chance that someone mistypes one of the constants so that it would still just choose something that worked. Is it worth throwing an error there instead or at least a warning that the constant was not recognized and LINEAR was chosen instead?

kjhollen replied:
ah oops I think I wasn't clear! any reason not to just have the else part (and lose the else-if)?

aferriss replied:
No reason at all except my oversight. Lets axe it!

kjhollen commented:
Thanks! This looks good to me. One thing I forgot to ask before, was if you can put an example in test/manual-test-examples/webgl ? There are a couple image assets in there already that you can use.

aferriss commented:
No problem, example added!

tcoppex commented:
Hi, 

So relative to issue #2189 I've been thinking a bit more about this and I'm not sure anymore if the two features could be easily merged.

As I see it, _p5.RendererGL.textureWrap_ is intended to change the value of the wrapping mode globally whereas _p5.Texture.setWrapMode_ change it locally, giving a more fine-grained control over the result, which is arguably better.

Ideally, here's what we want :


```javascript
function load_images() {
  A = loadImage('imageA.png');
  B = loadImage('imageB.png');
}

function draw() {
  // A wrapmode set to REPEAT.
  A.setTextureWrap(REPEAT, REPEAT);

  // Global wrapmode set to MIRROR.
  textureWrap(MIRROR);

  // A and B are MIRROR.
  textured_quad(A, u, v);
  textured_quad(B, u, v);

  translate(3, 0, 0);

  // Change A wrapmode to REPEAT.
  A.setTextureWrap(REPEAT, REPEAT);

  // A is REPEAT, B is MIRROR.
  textured_quad(A, u, v);
  textured_quad(B, u, v);
}
```

My first implementation matches the one from @aferriss  but the two proved incompatible.
I've tried somes angles, mainly storing states locally or globally, without much success.

Ideally _textureWrap()_ should broadcast its new state immediately to existing images,
which might be done by storing them in a buffer inside the p5 context when _createImage()_ is called
but this might broke the lifecycle intended by the user altogether.

Hopefully there is something simple I don't see, anyways any helps is welcome.

Thanks !

aferriss commented:
Hi @tcoppex ! Thanks for looking into this. Am I missing something or can't you just loop over the textures array stored in the canvas. Here is a little example I implemented on top of this PR that does exactly what you suggest (though I just implemented the global function in this sketch)

````
var A;
var B;
var Atex;

let canvas;
function preload() {
  A = loadImage('../../assets/UV_Grid_Sm.jpg');
  B = loadImage('../../assets/cat.jpg');
}

function setup() {
  canvas = createCanvas(windowWidth, windowHeight, WEBGL);
  Atex = canvas.getTexture(A);
}

function draw() {
  background(255);

  const dX = mouseX / windowWidth;
  const dY = mouseY / windowHeight;

  let u = lerp(1.0, 8.0, dX);
  let v = lerp(1.0, 8.0, dY);

  scale(width/9.0);

  // set just A to repeat
  Atex.setWrapMode(REPEAT, REPEAT);

  // set all textures to mirror
  textureWrap(MIRROR);

  push();
  translate(-1.5, -1, 0);
  textured_quad(A, u, v);
  pop();
  
  push();
  translate(-1.5, 1, 0);
  textured_quad(B, u, v);
  pop();

  // set just A to repeat
  Atex.setWrapMode(REPEAT, REPEAT);

  push();
  translate(1.5, -1, 0);
  textured_quad(A, u, v);
  pop();

  // b is still mirror
  push();
  translate(1.5, 1, 0);
  textured_quad(B, u, v);
  pop();
  
}

function textured_quad(tex, u, v) {
  texture(tex);
  beginShape(TRIANGLES);
  vertex(-1, -1,  0, 0, 0);
  vertex( 1, -1,  0, u, 0);
  vertex( 1,  1,  0, u, v);

  vertex( 1,  1,  0, u, v);
  vertex(-1,  1,  0, 0, v);
  vertex(-1, -1,  0, 0, 0);
  endShape();
}

function textureWrap(mode){
  for(let i =0; i<canvas.textures.length; i++){
    canvas.textures[i].setWrapMode(mode, mode);
  }
}
````

tcoppex commented:
Hey @aferriss, 
I did not saw that array, that seems perfect indeed, thanks.
Maybe you can go ahead and implement the feature along your PR if you're feeling like it ? it might be simpler that way.

Cheers :+1:

aferriss commented:
@tcoppex sure, I'll throw it in there too! Do you think it's good to add the method so that it can take either one or two arguments for maximum flexibility? For example...

````
// set clamp on both x and y axes
textureWrap(CLAMP); 

// set clamp on x axis and repeat on y axis
textureWrap(CLAMP, REPEAT);
```

tcoppex commented:
I think it won't hurt : it keeps the legacy with Processing and add flexibility :-)
(maybe @kjhollen can help though, I don't know if there is any philosophy behind multiple definitions in p5.js)
